src/Assistant.ts

```diff
@@ -0,0 +1,376 @@
+import {
+  AssistantThreadsSetStatusResponse,
+  AssistantThreadsSetSuggestedPromptsResponse,
+  AssistantThreadsSetTitleResponse,
+  ChatPostMessageArguments,
+} from '@slack/web-api';
+import processMiddleware from './middleware/process';
+import {
+  AllMiddlewareArgs,
+  AnyMiddlewareArgs,
+  Middleware,
+  SayFn,
+  SlackEventMiddlewareArgs,
+} from './types';
+import { AssistantInitializationError, AssistantMissingPropertyError } from './errors';
+import {
+  AssistantThreadContext,
+  AssistantThreadContextStore,
+  DefaultThreadContextStore,
+  GetThreadContextFn,
+  SaveThreadContextFn,
+} from './AssistantThreadContextStore';
+
+/**
+ * Configuration object used to instantiate the Assistant
+ */
+export interface AssistantConfig {
+  threadContextStore?: AssistantThreadContextStore;
+  threadStarted: AssistantThreadStartedMiddleware | AssistantThreadStartedMiddleware[];
+  threadContextChanged?: AssistantThreadContextChangedMiddleware | AssistantThreadContextChangedMiddleware[];
+  userMessage: AssistantUserMessageMiddleware | AssistantUserMessageMiddleware[];
+}
+
+/**
+ * Callback utilities
+ */
+interface AssistantUtilityArgs {
+  getThreadContext: GetThreadContextFn;
+  saveThreadContext: SaveThreadContextFn;
+  say: SayFn;
+  setStatus: SetStatusFn;
+  setSuggestedPrompts: SetSuggestedPromptsFn;
+  setTitle: SetTitleFn;
+}
+
+export interface SetStatusFn {
```

Comment:
This and the next three `Set*` interfaces are only used within this file. Can we remove the `export` on them, then? I want to limit as much as possible what we expose to developers, to try to limit any future breaking changes should those be necessary. It also helps keep track of what is used and what isn't.

Reply:
Sure! I followed what we did previously with similar implementations, like `WorkflowStep`. And I think we exported them after we released because developers were asking for it. We can not export these when we release it, if you'd like.

Reply:
If you are OK with it, I'd rather wait for devs to ask for it than to do it ahead of time, personally. When everything is exported, it makes it harder to make changes later. I've also already ran into situations where exported interfaces and bits were unused in the codebase entirely - but it's not obvious because tooling like linters don't raise that as an issue due to the `export` keyword.